src/sphinxcontrib/towncrier/__init__.py

```diff
@@ -71,11 +71,18 @@ def _get_changelog_draft_entries(
         # https://github.com/twisted/towncrier/pull/157#issuecomment-666549246
         # https://github.com/twisted/towncrier/issues/269
         extra_cli_args += '--config', str(config_path)
-    towncrier_output = subprocess.check_output(  # noqa: S603
-        TOWNCRIER_DRAFT_CMD + extra_cli_args,
-        cwd=str(working_dir) if working_dir else None,
-        universal_newlines=True,  # this arg has "text" alias since Python 3.7
-    ).strip()
+
+    try:
+        towncrier_output = subprocess.check_output(  # noqa: S603
+            TOWNCRIER_DRAFT_CMD + extra_cli_args,
+            cwd=str(working_dir) if working_dir else None,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,  # a "text" alias exists since Python 3.7
+        ).strip()
+
+    except subprocess.CalledProcessError as proc_err:
+        sys.stderr.write(proc_err.stderr.decode(sys.getfilesystemencoding()))
```

Comment:
I don't think that we should:
1) Put this specifically into stderr
2) Use a low-level stream interface without flushing

I just realized that reraising this exception is an abstraction leak. How about turning this into a `RuntimeError`?
Then, here https://github.com/sphinx-contrib/sphinxcontrib-towncrier/blob/532e343/src/sphinxcontrib/towncrier/__init__.py#L237-L238, it'd be catching that `RuntimeError`. And in this function it'd be `raise RuntimeError(subprocess_failure_detail) from proc_err`. We'd then construct a `subprocess_failure_detail` message on this level, but wouldn't interact with the low-level I/O letting Sphinx print out the error when it's appropriate...

@davfsa WDYT?

Reply:
Initially I was going to go with raising a RuntimeError, but thought that it would obscure a bit what the actual error was, but I have not tested it in full combination with sphinx these days due to not having the means to. I'll test out how it interacts with sphinx and change accordingly.

On your on `sys.syderr`, im using it as an alternative to `print`, but using output streams is still a bit new to me and things like flushing I have not really fully understood yet, so I'll follow your ideas there.

Furthermore, we don't really know why towncrier failed, we just know it did. The way I was testing this while coding it was by providing a path with a missing config file. This would cause it to exit with error code 1 while printing out 2 lines to sdterr: one with the actual error ("Configuration file not found") and one with extra information ("Looked under: <path>"). Translating that to a RuntimeError reason would be trivial, as I am not sure what all the possible ways towncrier has of failing.

Regardless, I'll have a look at how sphinx handles errors in extensions and go on from there. Thank you for pointing me in the right direction!

Reply:
> Translating that to a RuntimeError reason would be trivial, as I am not sure what all the possible ways towncrier has of failing.

I'd just maybe construct a string along the lines of:
```python
f"""
Invoking Towncrier failed.

Return code: {proc_err.rc}

Standard output:
{proc_err.stdout}

Standard error:
{proc_err.stdout}
"""
```
And feed that to the `RuntimeError` constructor.

Reply:
Played around with the code for a bit and found a nice way to handle the error. This will now be the output on an error occurring (https://github.com/sphinx-contrib/sphinxcontrib-towncrier/pull/58/commits/c8b3e39ac51be586712103f9e133dd7ccd15e8cb):

```
Running Sphinx v5.1.1
loading pickled environment... done
[autosummary] generating autosummary for: changelog/index.md, index.md
myst v0.18.0: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=[], disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, highlight_code_blocks=True, number_code_blocks=[], title_to_header=False, heading_anchors=3, heading_slug_func=None, footnote_transition=True, words_per_minute=200, sub_delimiters=('{', '}'), linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area')
building [mo]: targets for 0 po files that are out of date
building [dirhtml]: targets for 0 source files that are out of date
updating environment: 0 added, 2 changed, 0 removed
reading sources... [100%] index                                                                                                                                                                                                         
WARNING: Missing key 'towncrier' in file pyproject.toml
/home/davfsa/coding/hikari/docs/changelog/index.md:7: ERROR: Command '/home/davfsa/coding/hikari/.nox/sphinx/bin/python -m towncrier build --draft --version \\ Unreleased changes' exited with non-zero error code (1)

Standard output:
None

Standard error:
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/towncrier/__main__.py", line 3, in <module>
    cli()
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/towncrier/build.py", line 62, in _main
    return __main(
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/towncrier/build.py", line 88, in __main
    base_directory, config = load_config_from_options(directory, config_file)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/towncrier/_settings.py", line 45, in load_config_from_options
    config = load_config(base_directory)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/towncrier/_settings.py", line 74, in load_config
    return load_config_from_file(directory, config_file)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/towncrier/_settings.py", line 85, in load_config_from_file
    return parse_toml(directory, config)
  File "/home/davfsa/coding/hikari/.nox/sphinx/lib/python3.10/site-packages/towncrier/_settings.py", line 92, in parse_toml
    config = config["tool"]["towncrier"]
KeyError: 'towncrier'
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index                                                                                                                                                                                                          
generating indices... done
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 2 warnings.

The HTML pages are in public/docs/dirhtml.
```

I also realized this is a huge improvement from the old one, where it would just say "The command failed"

Reply:
@davfsa try not to mark conversations as resolved if you want people to actually see/find your replies. Finding the hidden ones is often a challenge on GitHub (that is if one remembers to do so...)